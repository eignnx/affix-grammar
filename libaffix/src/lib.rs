#[macro_use]
extern crate serde_derive;

pub mod fault;
pub mod gen;
pub mod literate;
pub mod parser;
pub mod semantics;

#[cfg(test)]
mod tests {
    use super::*;
    use std::convert::TryFrom;
    use std::fs;
    use std::path::PathBuf;

    static RNG_SEED: u64 = 123456;
    static MAX_VALUES: usize = 100;

    #[test]
    fn run_snapshots() {
        let inputs = PathBuf::from(std::env!("CARGO_MANIFEST_DIR"))
            .join("src")
            .join("snapshot_inputs");

        for dir_entry in fs::read_dir(inputs).unwrap() {
            let path = dir_entry.unwrap().path();
            let src_text = fs::read_to_string(&path).unwrap();
            let parse_res = parser::syntax::ParsedGrammar::try_from(&src_text[..]);

            let parse_snapshot_name = format!(
                "PARSE_SNAPSHOT__{}",
                path.file_stem().unwrap().to_string_lossy()
            );
            insta::assert_json_snapshot!(parse_snapshot_name, parse_res);

            if let Some(grammar) = parse_res.ok() {
                let generator = gen::Generator::new_seeded(grammar, RNG_SEED);
                let generated = generate_n(generator, MAX_VALUES);
                let gen_snapshot_name = format!(
                    "GEN_SNAPSHOT__{}",
                    path.file_stem().unwrap().to_string_lossy()
                );
                insta::assert_json_snapshot!(gen_snapshot_name, generated);
            }
        }
    }

    fn generate_n<R: rand::Rng>(
        mut gen: gen::Generator<R>,
        n: usize,
    ) -> fault::DynamicRes<Vec<String>> {
        let mut sentences = vec![];
        for _ in 0..n {
            match gen.generate() {
                Ok(sentence) => sentences.push(sentence),
                Err(fault::DynamicErr::MaxTrialsExceeded { .. }) => return Ok(sentences),
                Err(e) => return Err(e),
            }
        }
        Ok(sentences)
    }
}
