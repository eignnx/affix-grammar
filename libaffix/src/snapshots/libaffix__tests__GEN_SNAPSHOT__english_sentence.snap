---
source: libaffix/src/lib.rs
expression: generated
---
{
  "Ok": [
    "the marmot hits an obscene amount of tuna with a bat.",
    "the dog stumbles upon the chair during their annual checkup.",
    "the cow stumbles upon the couch while mother is out.",
    "the dog hits a lifetime supply of medical bills with a bat.",
    "the marmot stumbles upon a steaming pile of ketchup and mustard under the cover of darkness.",
    "look at that dog go. go dog, go.",
    "look at that cow go. go cow, go.",
    "the marmot stumbles upon a steaming pile of bees under the cover of darkness.",
    "the dog stumbles upon the chair under the cover of darkness.",
    "look at that marmot go. go marmot, go.",
    "the cow stumbles upon a steaming pile of medical bills upon stepping outdoors.",
    "the marmot stumbles upon a steaming pile of fruit-flies upon stepping outdoors. in the blink of an eye, the marmot hits an obscene amount of medical bills with a bat.",
    "the cow eats the couch.",
    "the dog stumbles upon a lifetime supply of caramel while on their morning walk.",
    "the marmot began setting an extremely loud marmot on fire.",
    "the dog stumbles upon an obscene amount of caramel during their annual checkup.",
    "the dog hits the dog's lunch with a bat.",
    "the dog stumbles upon the dog's lunch while on their morning walk.",
    "the dog began setting an extremely loud dog on fire.",
    "the dog hits an obscene amount of ketchup and mustard with a bat.",
    "the dog stumbles upon a steaming pile of jelly under the cover of darkness.",
    "the cow eats a lifetime supply of fruit-flies.",
    "the cow eats an extremely loud cow.",
    "the cow began setting an extremely loud cow on fire.",
    "the dog began setting a steaming pile of jelly on fire.",
    "the marmot stumbles upon a steaming pile of fruit-flies during their annual checkup.",
    "the dog began setting the couch on fire. but later that day, the dog stumbles upon the couch during their annual checkup. in the blink of an eye, the dog began setting the dog's lunch on fire.",
    "the marmot began setting a lifetime supply of caramel on fire.",
    "the marmot began setting the chair on fire.",
    "the cow eats a steaming pile of jelly. but later that day, look at that dog go. go dog, go. having completed the task at hand, the cow hits the couch with a bat.",
    "the dog began setting an obscene amount of tuna on fire.",
    "the dog began setting a lifetime supply of bees on fire.",
    "the marmot hits the couch with a bat.",
    "the dog began setting a lifetime supply of tuna on fire. having completed the task at hand, the dog began setting the chair on fire.",
    "the marmot hits the marmot's lunch with a bat.",
    "the dog stumbles upon an obscene amount of tuna while on their morning walk.",
    "the dog began setting the couch on fire.",
    "the dog began setting a lifetime supply of ketchup and mustard on fire.",
    "the cow eats a steaming pile of bees.",
    "the dog eats the couch.",
    "the marmot began setting an obscene amount of fruit-flies on fire.",
    "the marmot stumbles upon a lifetime supply of caramel upon stepping outdoors.",
    "the dog hits an extremely loud dog with a bat.",
    "the cow hits the couch with a bat.",
    "the dog eats an extremely loud dog.",
    "the cow hits an obscene amount of caramel with a bat.",
    "the marmot eats the couch.",
    "the marmot began setting the couch on fire.",
    "the cow stumbles upon a lifetime supply of caramel upon stepping outdoors.",
    "the dog hits a lifetime supply of ketchup and mustard with a bat.",
    "the cow stumbles upon a steaming pile of ketchup and mustard upon stepping outdoors. with a twitch of it's nose, the cow began setting the couch on fire.",
    "the dog stumbles upon an obscene amount of jelly while mother is out.",
    "the marmot eats an extremely loud marmot. with a twitch of it's nose, the marmot eats the couch.",
    "the dog eats the dog's lunch.",
    "the cow eats an obscene amount of tuna.",
    "the dog began setting a steaming pile of bees on fire.",
    "look at that cow go. go cow, go. with the greatest of ease, the marmot eats the marmot's lunch.",
    "the cow hits an obscene amount of tuna with a bat.",
    "the marmot eats an extremely loud marmot.",
    "the dog hits a lifetime supply of tuna with a bat.",
    "look at that cow go. go cow, go. but later that day, the cow hits the chair with a bat. with a twitch of it's nose, the cow hits a steaming pile of medical bills with a bat.",
    "the marmot eats an extremely loud marmot. with the greatest of ease, the marmot eats a lifetime supply of tuna. in the blink of an eye, the marmot hits an extremely loud marmot with a bat.",
    "look at that cow go. go cow, go. but later that day, the marmot hits the marmot's lunch with a bat.",
    "the dog hits the chair with a bat. having completed the task at hand, the dog began setting the couch on fire.",
    "the dog eats a steaming pile of bees.",
    "the cow stumbles upon a steaming pile of jelly under the cover of darkness.",
    "the dog stumbles upon a lifetime supply of jelly while on their morning walk.",
    "the cow eats an extremely loud cow. in the blink of an eye, the cow hits an obscene amount of fruit-flies with a bat. with the greatest of ease, the cow began setting the cow's lunch on fire. in the blink of an eye, look at that dog go. go dog, go.",
    "the cow stumbles upon the cow's lunch upon stepping outdoors.",
    "the cow hits an extremely loud cow with a bat. in the blink of an eye, the cow began setting a steaming pile of jelly on fire.",
    "the cow began setting the chair on fire.",
    "the cow began setting a steaming pile of fruit-flies on fire.",
    "the cow stumbles upon a lifetime supply of fruit-flies while mother is out.",
    "the marmot eats the chair.",
    "the dog began setting an extremely loud dog on fire. in the blink of an eye, the dog stumbles upon a lifetime supply of ketchup and mustard at the corner store.",
    "the dog began setting an obscene amount of medical bills on fire.",
    "the cow stumbles upon the cow's lunch during their annual checkup.",
    "the marmot eats the couch. in the blink of an eye, the marmot hits the couch with a bat.",
    "the marmot stumbles upon the couch while on their morning walk.",
    "the cow began setting an obscene amount of bees on fire.",
    "the marmot stumbles upon the chair under the cover of darkness.",
    "the dog eats an obscene amount of medical bills.",
    "the dog began setting the chair on fire.",
    "the dog eats the chair.",
    "the dog stumbles upon the dog's lunch while mother is out.",
    "the marmot began setting a steaming pile of tuna on fire.",
    "the cow began setting the couch on fire.",
    "the marmot eats a lifetime supply of medical bills.",
    "look at that dog go. go dog, go. with a twitch of it's nose, the dog began setting the couch on fire.",
    "look at that marmot go. go marmot, go. with a twitch of it's nose, look at that marmot go. go marmot, go.",
    "the dog hits the couch with a bat.",
    "the cow began setting a steaming pile of ketchup and mustard on fire.",
    "the marmot stumbles upon the marmot's lunch while mother is out.",
    "the marmot began setting a steaming pile of ketchup and mustard on fire.",
    "the dog eats a steaming pile of jelly.",
    "the dog stumbles upon the chair at the corner store.",
    "look at that dog go. go dog, go. with a twitch of it's nose, the dog eats the couch.",
    "the marmot began setting an obscene amount of tuna on fire.",
    "the marmot began setting an obscene amount of caramel on fire.",
    "the marmot began setting the couch on fire. with the greatest of ease, the marmot eats an extremely loud marmot."
  ]
}
